## Categorical variables

Categorical variables refer to variables whose range of values does not represent a meaningful continuous progression. Let's take the example of climate. One way of discussing climate might be to provide an average annual temperature for every city in our dataset. This represents a continuous value where, for example, it is meaningful that 8.1 is larger than 7.9 (in this case, it means hotter on average). It should be noted that this does not paint the whole picture, however. For instance, city A might have a higher average annual temperature than city B, but this does not mean that any given day is more likely to be hotter. For instance, city A might have very intense temperature fluctuations where it experiences both hotter and colder weather than city B. However, for the sake of simplicity, we can handle this value as a continuous one. On the other hand, each city on our dataset might be assigned a number for the climate type it has; 1 might refer to a hot dry summer climate, while 2 refers to a cold dry winter climate (Beirut and Beijing, respectively). In this case, it is not useful to take a city in between those two, Karachi for instance, and impute its climate type value to be 1.5, as this is meaningless. What would be more helpful is consider cities in a specific radius around Karachi and consider the mode (most frequent) climate category in that region. This is very likely to be Karachi's climate type as well.

In our dataset, we have some continuous variables, like 'price' and 'square footage', and categorical variables like 'view' and 'waterfront'. With regards to the examples discussed in class, 'bathrooms' and 'bedrooms', I tend to consider them a bit of both. As a real estate agent, I would not tell a client that their house, based on price, is likely to have 3.829 bathrooms (in this case, aggregating values and determining which category the house is likely to belong to is more useful.). However, if I were doing a broader study to correlate number of bathrooms with price, then I would find it permissible to treat bathroom count as a continuous variable. 

## Handling missing values

In [755]:
import pandas as pd

In [756]:
df = pd.read_csv('kc_house_data.csv')

In [757]:
df_ = df[['id', 'price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot']]

In [758]:
df_.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           21613 non-null  int64  
 1   price        21613 non-null  float64
 2   bedrooms     21613 non-null  int64  
 3   bathrooms    21613 non-null  float64
 4   sqft_living  21613 non-null  int64  
 5   sqft_lot     21613 non-null  int64  
dtypes: float64(2), int64(4)
memory usage: 1013.2 KB


In [759]:
import numpy as np

As I still haven't full understood the percentage method, I have chosen to introduce np.nan values manually based on conditions. I will read more about the percentage method later since it allows to introduce missing values at random.

In [760]:
df_.loc[df['bedrooms'] > 4.0, 'bedrooms'] = np.nan
df_.loc[df['bathrooms'] > 2.0, 'bathrooms'] = np.nan

In [761]:
df_.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           21613 non-null  int64  
 1   price        21613 non-null  float64
 2   bedrooms     19678 non-null  float64
 3   bathrooms    10371 non-null  float64
 4   sqft_living  21613 non-null  int64  
 5   sqft_lot     21613 non-null  int64  
dtypes: float64(3), int64(3)
memory usage: 1013.2 KB


The first method to deal with missing values is to delete those rows, using dropna(). First, we drop all rows with any missing value.

In [762]:
df_cleaned1 = df_.dropna()

In [763]:
df_cleaned1.info()
len(df_cleaned1)

<class 'pandas.core.frame.DataFrame'>
Index: 9979 entries, 0 to 21612
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           9979 non-null   int64  
 1   price        9979 non-null   float64
 2   bedrooms     9979 non-null   float64
 3   bathrooms    9979 non-null   float64
 4   sqft_living  9979 non-null   int64  
 5   sqft_lot     9979 non-null   int64  
dtypes: float64(3), int64(3)
memory usage: 545.7 KB


9979

This has reduced the count to 9979. However, I can also choose to drop just the rows with a missing 'bathroom' value, keeping the missing 'bedroom' values.

In [764]:
df_cleaned2 = df_.dropna(subset=['bathrooms'])

In [765]:
df_cleaned2.info()
len(df_cleaned2)

<class 'pandas.core.frame.DataFrame'>
Index: 10371 entries, 0 to 21612
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           10371 non-null  int64  
 1   price        10371 non-null  float64
 2   bedrooms     9979 non-null   float64
 3   bathrooms    10371 non-null  float64
 4   sqft_living  10371 non-null  int64  
 5   sqft_lot     10371 non-null  int64  
dtypes: float64(3), int64(3)
memory usage: 567.2 KB


10371

All rows with a missing 'bathroom' value have been dropped. In the 'bedrooms' column, there are additional missing values which we have not dropped.

Another way to deal with missing values is to replace them with a different value that makes sense for that variable type. In this case, I will fill missing 'bedroom' and 'bathroom' values with the respective modes.

In [766]:
df_.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           21613 non-null  int64  
 1   price        21613 non-null  float64
 2   bedrooms     19678 non-null  float64
 3   bathrooms    10371 non-null  float64
 4   sqft_living  21613 non-null  int64  
 5   sqft_lot     21613 non-null  int64  
dtypes: float64(3), int64(3)
memory usage: 1013.2 KB


In [767]:
df_['bedrooms'].mode()

0    3.0
Name: bedrooms, dtype: float64

In [768]:
bedroom_mode = df_['bedrooms'].mode()[0]

In [769]:
df_filled = df_.fillna(value={'bedrooms': bedroom_mode})

In [770]:
df_filled.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           21613 non-null  int64  
 1   price        21613 non-null  float64
 2   bedrooms     21613 non-null  float64
 3   bathrooms    10371 non-null  float64
 4   sqft_living  21613 non-null  int64  
 5   sqft_lot     21613 non-null  int64  
dtypes: float64(3), int64(3)
memory usage: 1013.2 KB


Missing bedroom values have been replaced. I will do the same for bathrooms, but instead of declaring a new variable for the filled table, I will specify inplace=True in the fillna function argument.

In [771]:
bathroom_mode = df_['bathrooms'].mode()[0]

In [772]:
df_filled.fillna(value={'bathrooms': bathroom_mode}, inplace=True)

In [773]:
df_filled.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           21613 non-null  int64  
 1   price        21613 non-null  float64
 2   bedrooms     21613 non-null  float64
 3   bathrooms    21613 non-null  float64
 4   sqft_living  21613 non-null  int64  
 5   sqft_lot     21613 non-null  int64  
dtypes: float64(3), int64(3)
memory usage: 1013.2 KB


All missing values have been replaced.